pulpcore/pulpcore/app/viewsets/repository.py

```diff
@@ -177,6 +186,65 @@ def destroy(self, request, repository_pk, number):
         )
         return OperationPostponedResponse([async_result], request)
 
+    def get_content_pk(self, url):
+        """
+        Extracts the Content's PK from a URL
+
+        This method also validates that the URL belongs to an existing Content resource
+
+        Args:
+            url (str): href that needs validation
+
+        Returns:
+            PK of the Content
+
+        Raises:
+            :class:`rest_framework.exceptions.ValidationError`: When the URL is not a valid
+                Content resource
+        """
+
+        try:
+            resolver = resolve(urlparse(url).path)
+        except Resolver404 as e:
+            raise ValidationError(detail=_("Invalid hyperlink. %s") % url)
+
+        if not resolver.view_name.startswith('content') or not resolver.view_name.endswith(
+                'detail'):
+            # URL exists, but does not correspond to Content resource
+            raise ValidationError(detail=_("Invalid hyperlink. %s") % url)
+
+        content_field = serializers.HyperlinkedRelatedField(view_name=resolver.view_name,
+                                                            queryset=Content.objects.all(),
+                                                            source='*', initial=url)
+        try:
+            content_field.run_validation(data=url)
+        except ValidationError as e:
+            # Append the URL of missing Content to the error message
+            e.detail[0] = "%s %s" % (e.detail[0], url)
+            raise e
+        return resolver.kwargs['pk']
+
+    def create(self, request, repository_pk):
+        add_content_units = []
+        remove_content_units = []
+        if 'add_content_units' in request.data:
+            for url in request.data['add_content_units']:
+                add_content_units.append(self.get_content_pk(url))
+
+        if 'remove_content_units' in request.data:
+            for url in request.data['remove_content_units']:
+                remove_content_units.append(self.get_content_pk(url))
+
+        result = tasks.repository.add_or_remove.apply_async_with_reservation(
+            tags.RESOURCE_REPOSITORY_TYPE, repository_pk,
+            kwargs={
+                'repository_pk': repository_pk,
+                'add_content_units': add_content_units,
```

Comment:
Could we just pass request.data.get('add_content_units')? That way all the url wrangling can happen in task land

Reply:
I was originally passing the URLs to the task, and I discovered that I was doing the same thing in the viewset and the task. Since I already validated that the URLs map to content units, I think it's ok to just let the database handle anything that goes missing between the REST API call and when the task executes.